jelares/portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -14,19 +14,99 @@
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
+  // Make the comments list data type
+  final ExampleComments comments = new ExampleComments(new ArrayList<String>());
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.setContentType("text/html;");
-    response.getWriter().println("<h1>Hello Jesus!</h1>");
+    // Read the query string to get comment limit
+    int commentLimit = Integer.parseInt(request.getParameter("num-comments"));
+
+    // Query to find all comment entities sorted from newest to oldest
+    Query query = new Query("Comment").addSort("timestamp", SortDirection.ASCENDING);
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+
+    // adding all comment entities to an ExampleComments instance
+    final ExampleComments comments = new ExampleComments(new ArrayList<String>());
+    for (Entity entity : results.asIterable(FetchOptions.Builder.withLimit(commentLimit))) {
+      long id = entity.getKey().getId();
+      String commentText = (String) entity.getProperty("comment-text");
+      comments.addComment(commentText);
+    }
+
+    response.setContentType("application/json;");
+    response.getWriter().println(commentsToJson(comments));
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Get the input from the form.
+    final String comment = request.getParameter("comment");
+    final long timestamp = System.currentTimeMillis();
+
+    // Add the input to the comments object
+    comments.addComment(comment);
+
+    // Add the input to datastore
+    Entity commentEntity = new Entity("Comment");
+    commentEntity.setProperty("comment-text", comment);
+    commentEntity.setProperty("timestamp", timestamp);
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    datastore.put(commentEntity);
+
+    // Redirect back to the HTML page.
+    response.sendRedirect("/comments/dataPage.html");
+  }
+
+  /** Turns ExampleComments object into JSON */
+  public String commentsToJson(ExampleComments comments) {
+    final int numberOfComments = comments.getNumComments();
+
+    // Initially there will be no comments
+    if (numberOfComments == 0) {
+      return "{}";
+
+    } else {
+      String jsonComments = "{";
+
+      /*
+       * Adds all comments to the jsonObject in the form:
+       * {"comment1": "text", "comment2": "text", ..., "commentN": "text"}
+       */
+      for (int commentNumber = 0; commentNumber < numberOfComments; commentNumber++) {
+        jsonComments += "\"comment" + commentNumber + "\": ";
+        jsonComments += "\"" + comments.getComment(commentNumber) + "\"";
+
+        if (commentNumber < (numberOfComments - 1)) {
+          jsonComments += ", ";
+        } else {
+          jsonComments += "}";
+        }
```

Comment:
Is there a way to do this using an existing library like Gson? Advantages of using an existing lib:
1. already tested (you won't need to write tests to verify correctness of this),
2. reduces logic needed (ideally you'd be able to pass in a Map or key-value store to the library, and it would format it correctly for you).

Reply:
In the walkthrough it said to use Gson or to write our own encoding, I wrote this one myself to practice encoding into JSON. When I go back in to address comments I will modify this part entirely to use Gson, I definitely agree with you in that using preexisting libraries is the best practice.

Reply:
I see, thanks for pointing out that part about the instructions.

I wonder if there's any higher-level API that could be used to make the formatting easier to read (even if less performant), especially at this point, without unit tests, the ability to easily manually inspect for errors is helpful.

Perhaps `String.join`?
https://mkyong.com/java/java-how-to-join-list-string-with-commas/

```
ArrayList<String> commentItems = new ...;
for (... commentNumber) {
  String commentItem = <e.g. 1:"comment text">;
  commentItems.add(commentItem);
}
return "{" + String.join(",", commentItems) + "}";
```

Reply:
On second thought @ykjchen , I read through the gson documentation and didn't find any immediate way to serialize individual entries in an array with as separate attributes of a json object. Do you know any way I could use gson to replicate what my commentsToJson function does?

Reply:
I used Gson with some custom code to solve this

Reply:
Was the change pushed?

Reply:
I had forgotten to remove the function but I did in the new commit